### PR TITLE
executor, planner: fix collation for hash join building

### LINKED ISSUE #23693: new collation + merge join get wrong result
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

new collation: on (confirmed that this is needed to trigger this bug)

[clustered_index_go_c.tbl_2-schema.sql.txt](https://github.com/pingcap/tidb/files/6228373/clustered_index_go_c.tbl_2-schema.sql.txt)
[clustered_index_go_c.tbl_2.000000000.sql.txt](https://github.com/pingcap/tidb/files/6228378/clustered_index_go_c.tbl_2.000000000.sql.txt)
```sql
select /*+  MERGE_JOIN( tbl_2 , tbl_2 ) */ tbl_2.* from tbl_2 where col_13 in ( select col_10 from tbl_2 where tbl_2.col_13 in ( 'Vftbh' , 'YjwPNaVCAJtKs' , 'DXMcdpfWCs' , 'peiJPqRH' , 'KInBGfejejv' , 'OFZi' , 'MBlDV' ) ) order by col_10 , col_11 , col_12 , col_13 , col_14 limit 901 ;
```
Sometimes the bug might not appear because the query plan is unstable.

### 2. What did you expect to see? (Required)
```
> select tbl_2.* from tbl_2 where col_13 in ( select col_10 from tbl_2 where tbl_2.col_13 in ( 'Vftbh' , 'YjwPNaVCAJtKs' , 'DXMcdpfWCs' , 'peiJPqRH' , 'KInBGfejejv' , 'OFZi' , 'MBlDV' ) ) order by col_10 , col_11 , col_12 , col_13 , col_14 limit 901 ;
+---------------+----------+--------------+--------+--------+
| col_10        | col_11   | col_12       | col_13 | col_14 |
+---------------+----------+--------------+--------+--------+
| h             |  3.04000 | HnyTNtGOGQcq | E      |    239 |
| zVImrWKTcuSlk | 43.37300 | LTwNDBR      | E      |     88 |
+---------------+----------+--------------+--------+--------+
2 rows in set (0.034 sec)

```

### 3. What did you see instead (Required)
```
> select /*+  MERGE_JOIN( tbl_2 , tbl_2 ) */ tbl_2.* from tbl_2 where col_13 in ( select col_10 from tbl_2 where tbl_2.col_13 in ( 'Vftbh' , 'YjwPNaVCAJtKs' , 'DXMcdpfWCs' , 'peiJPqRH' , 'KInBGfejejv' , 'OFZi' , 'MBlDV' ) ) order by col_10 , col_11 , col_12 , col_13 , col_14 limit 901 ;
Empty set, 1 warning (0.034 sec)
```

### 4. What is your TiDB version? (Required)

master

### PR DESCRIPTION
**Pull request by wjhuang2016**
### What problem does this PR solve?

Issue Number: close #23693 

Problem Summary:
outerHashKeys[i].GetType().Charset, outerHashKeys[i].GetType().Collate = chs, coll

collate is modified here, when building the merge join, the collate is wrong.

### What is changed and how it works?

Correct the collation in the executor.

### Related changes

- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test

Side effects

### Release note 

- fix collation for hash join building

### CODE REVIEW

File: executor/builder.go [hunk 1]
```diff
@@ -2358,6 +2358,8 @@ func (b *executorBuilder) buildIndexLookUpJoin(v *plannercore.PhysicalIndexJoin)
 		if innerTypes[i].EvalType() == types.ETString {
 			innerTypes[i].Flen = types.UnspecifiedLength
 		}
+		// Use the probe table's collation.
+		outerTypes[i].Collate = innerTypes[i].Collate
```
Comment (XuHuaiyu) [inline executor/builder.go:2362]: Do we need to deriveCollation here?

Comment (wjhuang2016) [reply executor/builder.go:2362]: No, we must use the probe table's collation.

Comment (wjhuang2016) [reply executor/builder.go:2362]: By the way, it causes Panic: index out of range.

**Comment by XuHuaiyu** [comment]
We need to modify  `outerCtx.rowType[hashCol[i]].collate `

**Comment by XuHuaiyu** [comment]
This test need to add a `drop table if exists`
``` sql
[2021-04-01T09:21:18.951Z] FAIL: integration_test.go:9003: testIntegrationSerialSuite.TestCollationPrefixClusteredIndex
[2021-04-01T09:21:18.951Z] 
[2021-04-01T09:21:18.951Z] integration_test.go:9009:
[2021-04-01T09:21:18.951Z]     tk.MustExec("create table t (k char(20), v int, primary key (k(4)) clustered, key (k)) collate utf8mb4_general_ci;")
[2021-04-01T09:21:18.951Z] /home/jenkins/agent/workspace/tidb_ghpr_check_2/go/src/github.com/pingcap/tidb/util/testkit/testkit.go:217:
[2021-04-01T09:21:18.951Z]     tk.c.Assert(err, check.IsNil, check.Commentf("sql:%s, %v, error stack %v", sql, args, errors.ErrorStack(err)))
[2021-04-01T09:21:18.951Z] ... value *errors.withStack = [schema:1050]Table 'test.t' already exists ("[schema:1050]Table 'test.t' already exists")
```

**Review by qw4990 — APPROVED** [review]

**Pull request merged**